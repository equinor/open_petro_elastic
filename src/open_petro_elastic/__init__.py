import importlib.metadata
from .float_vectorize import float_vectorize


__version__ = importlib.metadata.version(__package__ or __name__)

__all__ = [
    "float_vectorize",
]
