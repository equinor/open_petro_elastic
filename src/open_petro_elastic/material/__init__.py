"""
The materials submodule contains the models for materials used
for open_petro_elastic and can be used as a standalone library.

Sentral to the library is the Material class which carries the
elastic properties of any given material

    >>> from open_petro_elastic.material.sandstone import hertz_mindlin
    >>> from open_petro_elastic.material import Material
    >>> mineral = Material(
    ...        density=2.75,
    ...        bulk_modulus=20000,
    ...        shear_modulus=7,
    ... )
    >>> hertz_mindlin(mineral, 0.4, 1.0)
    Material(shear_modulus=3.8..., bulk_modulus=3.179...,...density=1.65)


"""
from .fluid import brie_fluid_mixing, fluid_substitution, wood_fluid_mixing
from .hashin_shtrikman import hashin_shtrikman_average, hashin_shtrikman_walpole
from .material import Material, polyval_material

__all__ = [
    "Material",
    "hashin_shtrikman_average",
    "hashin_shtrikman_walpole",
    "fluid_substitution",
    "brie_fluid_mixing",
    "wood_fluid_mixing",
    "polyval_material",
]
