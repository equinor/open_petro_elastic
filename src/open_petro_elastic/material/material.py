"""
The Material class is the data container for
elastic properties of one material.
"""
import numpy as np
from numpy.polynomial.polynomial import polyval2d

from .seismic_velocity import (
    calculate_bulk_modulus,
    calculate_primary_velocity,
    calculate_secondary_velocity,
    calculate_shear_modulus,
)


def polyval_material(
    value,
    material,
    density_coefficients,
    primary_coefficients,
    secondary_coefficients,
    use_moduli=False,
):
    """
    :returns: New Material with properties adjusted to a 2d polynomial, i.e.

    new_density = sum( c[i][j] * old_density^i * value^j)

    :param value: One dimension of the polynomial.
    :param material: The material to adjust.
    :param density_coefficients: The coefficients of the polynomial for density
    :param primary_coefficients: The coefficients of the polynomial for the
        primary property (bulk_modulus/primary_velocity)
    :param secondary_coefficients: The coefficients of the polynomial for the
        secondary property (shear_modulus/secondary_velocity)
    :param use_moduli: Whether properties are bulk_modulus/shear_modulus or
        primary_velocity/secondary_velocity
    """

    @vectorize_material
    def fit(d, p, s, v):
        a = polyval2d(
            d,
            v,
            density_coefficients,
        )
        b = polyval2d(
            p,
            v,
            primary_coefficients,
        )
        c = polyval2d(
            s,
            v,
            secondary_coefficients,
        )
        if use_moduli:
            return Material(density=a, bulk_modulus=b, shear_modulus=c)
        else:
            return Material(density=a, primary_velocity=b, secondary_velocity=c)

    if use_moduli:
        return fit(
            material.density, material.bulk_modulus, material.shear_modulus, value
        )
    else:
        return fit(
            material.density,
            material.primary_velocity,
            material.secondary_velocity,
            value,
        )


def vectorize_material(f):
    """
    vectorize a function that returns materials so that each
    property member (density, velocity, moduli) are vectors
    of floats.
    """

    def fun(*args, **kwargs):
        vector_of_materials = np.vectorize(f, otypes=[Material])(*args, **kwargs)
        if isinstance(vector_of_materials, np.ndarray):
            if len(vector_of_materials.shape) > 0:
                return Material(
                    density=np.array([m.density for m in vector_of_materials]),
                    primary_velocity=np.array(
                        [m.primary_velocity for m in vector_of_materials]
                    ),
                    secondary_velocity=np.array(
                        [m.secondary_velocity for m in vector_of_materials]
                    ),
                )
            else:
                return vector_of_materials[()]
        else:
            return vector_of_materials

    return fun


class Material:
    """
    A isotropic material having bulk modulus,
    shear modulus, and density. These can be derived from, or
    be used to derive velocity of primary and secondary waves
    according to the usual equation for estimating primary
    and secondary wave velocities in isotropic media. See
    seismic_velocity.py
    """

    def __init__(
        self,
        bulk_modulus=None,
        shear_modulus=None,
        density=None,
        secondary_velocity=None,
        primary_velocity=None,
    ):
        """
        :param bulk_modulus: The bulk modulus of the material in Pa.
        :param shear_modulus: The shear modulus of the material in Pa.
        :param density: The density of the material in kg/m^3
        :param primary_velocity: The velocity of secondary waves
            in the material in m/s.
        :param secondary_velocity: The velocity of secondary waves
            in the material in m/s.
        :return: Mineral with given properties, properties not given
            will be calculated, see Mineral.
        """
        self.bulk_modulus = bulk_modulus
        self.shear_modulus = shear_modulus
        self.density = density
        self.primary_velocity = primary_velocity
        self.secondary_velocity = secondary_velocity
        self.set_defaults()

        if np.any(self.density < 0):
            raise ValueError("Material should not have negative density")
        if np.any(self.bulk_modulus < 0):
            raise ValueError("Material should not have negative bulk modulus")
        if np.any(self.shear_modulus < 0):
            raise ValueError("Material should not have negative shear modulus")
        if np.any(self.primary_velocity < 0):
            raise ValueError("Material should not have negative primary velocity")
        if np.any(self.secondary_velocity < 0):
            raise ValueError("Material should not have negative secondary velocity")
        if np.any(self.primary_velocity < ((4 / 3) ** 0.5) * self.secondary_velocity):
            raise ValueError(
                "primary velocity must be at least sqrt(4/3) times secondary velocity"
            )

    def set_defaults(self):
        if self.shear_modulus is None:
            if self.secondary_velocity is None or self.density is None:
                raise ValueError("Cannot calculate bulk modulus for material")
            self.shear_modulus = calculate_shear_modulus(
                self.secondary_velocity, self.density
            )
        if self.bulk_modulus is None:
            if self.primary_velocity is None or self.density is None:
                raise ValueError("Cannot calculate bulk modulus for material")
            self.bulk_modulus = calculate_bulk_modulus(
                self.primary_velocity, self.shear_modulus, self.density
            )
        if self.primary_velocity is None:
            if self.bulk_modulus is None or self.density is None:
                raise ValueError("Cannot calculate primary velocity for material")
            self.primary_velocity = calculate_primary_velocity(
                self.bulk_modulus, self.shear_modulus, self.density
            )
        if self.secondary_velocity is None:
            if self.shear_modulus is None or self.density is None:
                raise ValueError("Cannot calculate secondary velocity for material")
            self.secondary_velocity = calculate_secondary_velocity(
                self.shear_modulus, self.density
            )

    @property
    def poisson_ratio(self):
        """
        Calculates poissons ratio from bulk modulus and shear modulus using
        usual conversion formulas, see
        https://en.wikipedia.org/wiki/Template:Elastic_moduli and
        G. Mavko, T. Mukerji, J. Dvorkin. The Rock Physics Handbook. Cambridge
        University Press 2003 (paperback). ISBN 0-521-54344-4
        """
        return (3 * self.bulk_modulus - 2 * self.shear_modulus) / (
            2 * (3 * self.bulk_modulus + self.shear_modulus)
        )

    @property
    def primary_wave_modulus(self):
        """
        Calculates primary wave modulus from bulk modulus and shear modulus using
        usual conversion formulas, see
        https://en.wikipedia.org/wiki/Template:Elastic_moduli and
        G. Mavko, T. Mukerji, J. Dvorkin. The Rock Physics Handbook. Cambridge
        University Press 2003 (paperback). ISBN 0-521-54344-4
        """
        return self.bulk_modulus + 4 / 3 * self.shear_modulus

    @property
    def acoustic_impedance(self):
        return self.primary_velocity * self.density

    @property
    def shear_impedance(self):
        return self.secondary_velocity * self.density

    def __repr__(self):
        return str(self)

    def __str__(self):
        return (
            "Material("
            + (
                ""
                if self.shear_modulus is None
                else f"shear_modulus={self.shear_modulus}, "
            )
            + (
                ""
                if self.bulk_modulus is None
                else f"bulk_modulus={self.bulk_modulus}, "
            )
            + (
                ""
                if self.primary_velocity is None
                else f"primary_velocity={self.primary_velocity}, "
            )
            + (
                ""
                if self.secondary_velocity is None
                else f"secondary_velocity={self.secondary_velocity}, "
            )
            + f"density={self.density})"
        )
