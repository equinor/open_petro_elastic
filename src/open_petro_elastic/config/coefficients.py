from typing import Optional, Union

try:
    from pydantic.v1.dataclasses import dataclass
except ImportError:
    from pydantic.dataclasses import dataclass

from open_petro_elastic.material import polyval_material

from .pydantic_config import PetroElasticConfig
from .vector_type import Array


@dataclass(config=PetroElasticConfig)
class ModuliCoefficients:
    density: Optional[Array[float]] = None
    bulk_modulus: Optional[Array[float]] = None
    shear_modulus: Optional[Array[float]] = None

    @property
    def as_list(self):
        return [self.density, self.bulk_modulus, self.shear_modulus]

    @property
    def use_moduli(self):
        return True

    def polyfit(self, value, material):
        return polyval_material(
            value,
            material,
            self.density,
            self.bulk_modulus,
            self.shear_modulus,
            use_moduli=True,
        )


@dataclass(config=PetroElasticConfig)
class VelocityCoefficients:
    density: Optional[Array[float]] = None
    secondary_velocity: Optional[Array[float]] = None
    primary_velocity: Optional[Array[float]] = None

    @property
    def as_list(self):
        return [self.density, self.primary_velocity, self.secondary_velocity]

    @property
    def use_moduli(self):
        return False

    def polyfit(self, value, material):
        return polyval_material(
            value,
            material,
            self.density,
            self.primary_velocity,
            self.secondary_velocity,
            use_moduli=False,
        )


@dataclass(config=PetroElasticConfig)
class VpOverVsCoefficients:
    density: Optional[Array[float]] = None
    vp_over_vs: Optional[Array[float]] = None
    bulk_modulus: Optional[Array[float]] = None

    @property
    def as_list(self):
        return [self.density, self.bulk_modulus, self.vp_over_vs]


PolyfitCoefficients = Union[ModuliCoefficients, VelocityCoefficients]
Coefficients = Union[VpOverVsCoefficients, ModuliCoefficients, VelocityCoefficients]
