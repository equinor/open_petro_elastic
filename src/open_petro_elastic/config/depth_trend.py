from typing import Optional

import numpy as np

try:
    from pydantic.v1.dataclasses import dataclass
except ImportError:
    from pydantic.dataclasses import dataclass

from typing_extensions import Literal

from .coefficients import PolyfitCoefficients
from .pydantic_config import PetroElasticConfig
from .vector_type import Array


@dataclass(config=PetroElasticConfig)
class DepthTrend:
    """
    Adjusts given dry rock to depth. Difference is
    calculated from given polynomial.
    :param reference_dry_rock: Dry rock to adjust at reference depth.
    :param depth: Difference in depth from reference.
    :param coefficients: The polynomial coefficients to used to adjust the material,
    such that the terms in the polynomial are p[i,j] * depth**i *
    mineral.bulk_modulus ** j.
    """

    depth: Array[float]
    coefficients: PolyfitCoefficients
    type: Literal["depth_trend"] = "depth_trend"
    reference_depth: Optional[Array[float]] = None
    max_depth: Optional[Array[float]] = None

    @property
    def effective_depth(self):
        result = np.asarray(self.depth).copy()
        if self.max_depth is not None:
            np.clip(result, None, self.max_depth)
        if self.reference_depth is not None:
            result = result - self.reference_depth
        return result

    def adjust_material(
        self,
        material,
        pressure=None,
        porosity=None,
        mineral=None,
    ):
        """
        :returns: The material at adjusted depth.
        """
        return self.coefficients.polyfit(
            self.effective_depth,
            material,
        )
