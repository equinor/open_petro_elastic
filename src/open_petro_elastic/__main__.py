import argparse
import sys
import warnings
from functools import wraps
from os import path
from shutil import copytree
from traceback import print_tb

import numpy as np
import pandas as pd
import yaml
from pydantic import parse_obj_as

import open_petro_elastic

from .config.input import ColumnInsertionError, Input
from .material import fluid_substitution


def warn_without_traceback(message, category, filename, lineno, file=None, line=None):

    log = file if hasattr(file, "write") else sys.stderr
    log.write(f"Warning: {message}\n")


def calibration_error(calculated, calibrated):
    return (calculated - calibrated) / calibrated


def print_calibration_error(name, error):
    print(
        f"""{name}:
min: {np.min(error)}
mean: {np.mean(error)}
max: {np.max(error)}
"""
    )


def calculate_results(inp):
    mixed_mineral = inp.minerals.as_mixture
    mixed_fluid = inp.fluids.as_mixture(inp.pressure)
    dry_rock_material = inp.dry_rock.material(mixed_mineral, inp.pressure)
    dense_packing = inp.dry_rock.nonporous(mixed_mineral)
    saturated_rock = fluid_substitution(
        dry_rock_material,
        dense_packing,
        mixed_fluid,
        inp.dry_rock.porosity,
    )
    index = [0]
    if inp.data is not None:
        index = inp.data.index
    return pd.DataFrame(
        {
            "ksat": saturated_rock.bulk_modulus,
            "kmin": mixed_mineral.bulk_modulus,
            "kdry": dry_rock_material.bulk_modulus,
            "mysat": saturated_rock.shear_modulus,
            "rsat": saturated_rock.density,
            "kmin_fls": dense_packing.bulk_modulus,
        },
        index=index,
    )


def takes_stream(i, mode):
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            if len(args) > i and args[i] is not None and isinstance(args[i], str):
                with open(args[i], mode) as f:
                    return func(*args[:i], f, *args[i + 1 :], **kwargs)
            else:
                return func(*args, **kwargs)

        return wrapper

    return decorator


class InputError(ValueError):
    pass


def load_config_from_yaml(config_file):
    try:
        config = yaml.safe_load(config_file)
    except yaml.YAMLError as exc:
        if hasattr(exc, "problem_mark"):
            m = "Error while parsing the config file:\n"
            if exc.context is not None:
                m += f"{exc.problem_mark}\n{exc.problem}\n{exc.context}"
            else:
                m += f"\n{exc.problem_mark}\n{exc.problem}"
        else:
            m = "Something went wrong while parsing the config file."
        raise InputError(m) from exc
    if not isinstance(config, dict):
        raise InputError("Unexpected format of config file")
    return config


def load_data_from_csv(data_file):
    try:
        if data_file is not None:
            data = pd.read_csv(data_file)
        else:
            data = None
    except pd.errors.ParserError as e:
        raise InputError("Error while reading data file: {e}") from e
    except pd.errors.EmptyDataError as e:
        raise InputError("Data file seems to be empty") from e
    return data


@takes_stream(0, "r")
def make_input(config_file, data_file):
    """
    :param config_file: Either path to or opened file to yaml config.
    :param data_file: Either path to or opened file to csv data file.
    :return: Input read from config and data
    """
    config = load_config_from_yaml(config_file)
    data = load_data_from_csv(data_file)
    try:
        config["data"] = data
        return parse_obj_as(Input, config)
    except ColumnInsertionError as e:
        raise InputError(f"Error while inserting column data from csv: {e}") from e
    except ValueError as e:
        raise InputError(f"Inconsistent values in input file: {e}") from e


def make_per_row_input(config_file, data_file):
    config = load_config_from_yaml(config_file)
    data = load_data_from_csv(data_file)
    inputs = []
    for i, row in data.iterrows():
        try:
            config["data"] = row.to_frame().transpose()
            inputs.append(parse_obj_as(Input, config))
        except ColumnInsertionError as e:
            print(f"Error while inserting column data from csv: {e}")
            inputs.append(None)
        except ValueError as e:
            print(f"Inconsistent values in input file for line {i}: {e}")
            inputs.append(None)
        except Exception as e:
            print(f"Encountered error while creating input for line {i}: {e}")
            inputs.append(None)
    return inputs


class CalibrationError(ValueError):
    pass


def run_calibration(calibration_file, results):
    try:
        calibration_data = pd.read_csv(calibration_file)
        for key in ["kmin", "kdry", "ksat", "kmin_fls"]:
            result = results[key]
            calibration = calibration_data[key]
            error = calibration_error(result, calibration)
            print_calibration_error(key, error)
    except pd.errors.ParserError as e:
        raise CalibrationError(f"Could not read calibration file: {e}") from e
    except pd.errors.EmptyDataError as e:
        raise CalibrationError("Calibration file seems to be empty") from e
    except (KeyError, IndexError) as e:
        raise CalibrationError(
            f"Error while getting column from calibration data: {e}"
        ) from e


def run_with_threshold(args):
    if args.data_file is None:
        print("In order to use success threshold, data must be given in csv file.")
        return -5
    inputs = make_per_row_input(args.config_file, args.data_file)
    num_failures = 0
    results = pd.DataFrame()
    for idx, inp in enumerate(inputs):
        if inp is None:
            num_failures += 1
            continue
        try:
            results = results.append(calculate_results(inp))
        except Exception as e:
            print(f"Encountered error while calculating rows {id}: {e}")
            num_failures += 1
    if len(inputs) == 0:
        success_percentage = 100
    else:
        success_percentage = (1 - num_failures / len(inputs)) * 100
    if success_percentage < args.success_threshold:
        print(
            f"Percentage of successes {success_percentage} did not reach threshold {args.success_threshold}"
        )
        return -5
    try:
        results.to_csv(args.output_file)
    except Exception as e:
        print(f"Encountered unexpected error while writing output: {e}")
        print_tb(e.__traceback__)
        return -1

    if args.calibration_data is not None:
        try:
            run_calibration(args.calibration_data, results)
        except CalibrationError as e:
            print(f"Error encountered in calibration: {e}")
            return -4
        except Exception as e:
            print(f"Encountered unexpected error during calibration: {e}")
            print_tb(e.__traceback__)
            return -1
    return 0


def run(args):
    if args.success_threshold is not None:
        return run_with_threshold(args)
    try:
        inp = make_input(args.config_file, args.data_file)
    except InputError as e:
        print(f"Could not read input: {e}")
        return -2
    except Exception as e:
        print(f"Encountered unexpected error while reading input: {e}")
        print_tb(e.__traceback__)
        return -1

    try:
        results = calculate_results(inp)
    except ValueError as e:
        print(f"Encountered invalid value in petro elastic calculation: {e}")
        return -3
    except Exception as e:
        print(f"Encountered unexpected error in petro elastic calculation: {e}")
        print_tb(e.__traceback__)
        return -1

    try:
        results.to_csv(args.output_file)
    except Exception as e:
        print(f"Encountered unexpected error while writing output: {e}")
        print_tb(e.__traceback__)
        return -1

    if args.calibration_data is not None:
        try:
            run_calibration(args.calibration_data, results)
        except CalibrationError as e:
            print(f"Error encountered in calibration: {e}")
            return -4
        except Exception as e:
            print(f"Encountered unexpected error during calibration: {e}")
            print_tb(e.__traceback__)
            return -1
    return 0


def generate_tutorial(out_dir):
    here = path.dirname(__file__)
    tutorial_path = path.join(here, "tutorial_config")
    copytree(tutorial_path, out_dir)


class GenerateTutorialAction(argparse.Action):
    def __call__(self, parser, namespace, values, option_string=None):
        try:
            generate_tutorial(values)
        except Exception as err:
            print(f"Could not write tutorial config: {err}")
            parser.exit(-4)
        parser.exit()


def percentage(value):
    try:
        value = float(value)
    except (ValueError, TypeError):
        raise argparse.ArgumentTypeError(f"Value must be float {value}")
    if 0.0 <= value <= 100.0:
        return value
    raise argparse.ArgumentTypeError(f"Value must be in range [0, 100] {value}")


TOOL_DESCRIPTION = """
The Equinor petro-elastic modelling tool. Models elastic properties
of sandstone saturated with oil, gas, and brine. See github.com/equinor/open_petro_elastic
for documentation.
"""


def make_parser(prog="open_petro_elastic"):
    ap = argparse.ArgumentParser(prog=prog, description=TOOL_DESCRIPTION)
    ap.add_argument(
        "config_file", type=argparse.FileType("r"), help="petro elastic yaml input file"
    )
    ap.add_argument(
        "--data-file",
        type=argparse.FileType("r"),
        dest="data_file",
        help="Optional additional csv file for data to be inserted into config.",
    )
    ap.add_argument(
        "--calibration",
        type=argparse.FileType("r"),
        dest="calibration_data",
        help="Optional calibration data, if given, an error statistic is calculated and printed.",
    )
    ap.add_argument(
        "--output_file",
        type=argparse.FileType("w"),
        default=sys.stdout,
        help="Optional output file, defaults to stdout.",
    )
    ap.add_argument(
        "--version", action="version", version=open_petro_elastic.__version__
    )
    ap.add_argument(
        "--generate-tutorial",
        action=GenerateTutorialAction,
        help="Write tutorial config. Takes directory to write to as argument.",
    )
    ap.add_argument(
        "--success_threshold",
        type=percentage,
        default=None,
        help="Optional percentage required for success."
        + " Default is 100 percent with no partial results given."
        + " Unsuccessful rows will be given NaN. Note that "
        + "this option can deteriorate performance.",
    )
    return ap


def parse_arguments(argv):
    ap = make_parser(argv[0])
    return ap.parse_args(argv[1:])


def main():
    warnings.showwarning = warn_without_traceback
    sys.exit(run(parse_arguments(sys.argv)))


if __name__ == "__main__":
    main()
