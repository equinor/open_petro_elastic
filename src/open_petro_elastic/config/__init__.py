"""
The open_petro_elastic.config module contains the internal representation of
the configuration file. The parsing of the configuration file goes through
three representations 1) YAML, 2) dictionary of yaml layout 3) Internal representation
as objects built from each section of the config file.

Each of the objects can be built from dictionaries, but it is possible to
construct directly. Generally, the constructor takes the unpacked
corresponding section of the dictionary, ie.  Fluids(**config["fluids"]) where
config is the dictionary from step 2.

"""

from .coefficients import Coefficients
from .constituent import Constituent
from .depth_trend import DepthTrend
from .dry_rock import DryRock
from .fluids import Fluids
from .minerals import Minerals
from .pressure import Pressure
from .pressure_dependency import PressureDependency

__all__ = [
    "DryRock",
    "Pressure",
    "Coefficients",
    "PressureDependency",
    "DepthTrend",
    "Minerals",
    "Constituent",
    "Fluids",
]
