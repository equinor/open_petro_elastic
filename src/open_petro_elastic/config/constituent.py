from typing import Optional

import numpy as np
from typing_extensions import Literal

try:
    from pydantic.v1 import validator
    from pydantic.v1.dataclasses import dataclass
except ImportError:
    from pydantic import validator
    from pydantic.dataclasses import dataclass

from open_petro_elastic.config.fluid_model_providers import fluid_model_providers
from open_petro_elastic.material import Material

from .material_type import MaterialType
from .pydantic_config import PetroElasticConfig
from .vector_type import Array


@dataclass(config=PetroElasticConfig)
class Constituent:
    """
    A Constituent is a part of a composite material to a
    volume or mass fraction.
    :param material: The constituent material which is part of a composite material.
    :param fraction: The ratio of constituent material to composite material.
    """

    material: MaterialType
    fraction: Optional[Array[float]] = None
    fluid_model: Optional[Literal[tuple(fluid_model_providers.keys())]] = None

    def as_material(
        self,
        temperature,
        pressure,
        fluid_model_provider=fluid_model_providers["default"],
    ):
        fluid_model_provider = (
            fluid_model_provider
            if self.fluid_model is None
            else fluid_model_providers[self.fluid_model]
        )
        return self.material.as_material(
            temperature, pressure, fluid_model_provider=fluid_model_provider
        )

    @validator("fraction")
    def fraction_is_ratio(cls, v):
        if v is None:
            return v
        if np.any(v < -1e-5):
            raise ValueError("Constituent fraction must be greater than zero")
        if np.any(v > 1 + 1e-5):
            raise ValueError("Constituent fraction must be less than one")
        return v

    @validator("material", pre=True)
    def wrap_a_material(cls, v):
        if isinstance(v, Material):
            return dict(
                density=v.density,
                bulk_modulus=v.bulk_modulus,
                shear_modulus=v.shear_modulus,
            )
        return v


def fix_one_fraction(constituents):
    """
    Given a list of all constituents of a composite material,
    one of which has fraction set to None, sets the value of
    that missing fraction so that the sum of fractions is 1.0.
    """
    are_none_fraction = [c for c in constituents if c.fraction is None]
    sum_rest = sum([c.fraction for c in constituents if c.fraction is not None])
    epsilon = 1e-5
    if np.any(sum_rest > 1.0 + epsilon):
        raise ValueError(
            f"Sum of constituent fractions {[c.fraction for c in constituents]} should not exceed 1.0, but is {sum_rest}"
        )
    if len(are_none_fraction) == 0:
        return
    if len(are_none_fraction) > 1:
        raise ValueError(
            f"Number of undefined fractions {[c.fraction for c in constituents]} should not be more than 1"
        )
    are_none_fraction[0].fraction = 1.0 - sum_rest
