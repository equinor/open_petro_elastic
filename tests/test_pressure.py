import numpy as np
import pytest
from numpy.testing import assert_allclose
from predicates import assert_similar_material

from open_petro_elastic.config import Pressure, PressureDependency
from open_petro_elastic.material import Material


def test_pressure_dependency_no_pressure_poly_fit(
    snapshot,
    dryrock_reference_material,
    overburden_pressure,
    reference_pressure,
    truncate_pressure,
    polynomial_fit_pressure_coefficients,
):
    presrock = 1.0
    pressure = Pressure(
        overburden=overburden_pressure,
        max_effective=truncate_pressure,
        reference=reference_pressure,
        rock=presrock,
        fluid=presrock,
    )
    snapshot.assert_match(
        PressureDependency(
            model={
                "type": "polyfit",
                "coefficients": polynomial_fit_pressure_coefficients,
            },
        ).adjust_material(dryrock_reference_material, pressure)
    )


def test_pressure_dependency_vector_pressure(dryrock_reference_material):
    pressure = Pressure(
        overburden=[20, 20],
        max_effective=[20, 20],
        reference=[5, 6],
        rock=[7, 8],
        fluid=[7, 8],
    )
    dryrock = PressureDependency(
        model={
            "type": "polyfit",
            "coefficients": {
                "bulk_modulus": [9, 10],
                "shear_modulus": [11, 12],
                "density": [13, 14],
            },
        },
    ).adjust_material(dryrock_reference_material, pressure)

    refb = dryrock_reference_material.bulk_modulus
    rockp = pressure.truncated_rock
    refp = pressure.truncated_reference
    assert_allclose(dryrock.bulk_modulus, (10 + 9 * rockp) / (10 + 9 * refp) * refb)


def test_pressure_dependency_empty_vector(dryrock_reference_material):
    pressure = Pressure(
        overburden=[],
        max_effective=[],
        reference=[],
        rock=[],
        fluid=[],
    )
    dryrock = PressureDependency(
        model={
            "type": "polyfit",
            "coefficients": {
                "bulk_modulus": [],
                "shear_modulus": [],
                "density": [],
            },
        },
    ).adjust_material(dryrock_reference_material, pressure)

    assert len(dryrock.bulk_modulus) == 0
    assert len(dryrock.density) == 0
    assert len(dryrock.shear_modulus) == 0


def test_pressure_dependency_no_pressure_power_fit(
    snapshot,
    dryrock_reference_material,
    overburden_pressure,
    reference_pressure,
    truncate_pressure,
):
    presrock = 1.0
    pressure = Pressure(
        overburden=overburden_pressure,
        max_effective=truncate_pressure,
        reference=reference_pressure,
        rock=presrock,
        fluid=presrock,
    )
    d = PressureDependency(
        model={
            "type": "logfit",
            "coefficients": {
                "density": np.ones(2),
                "primary_velocity": np.ones(2),
                "secondary_velocity": np.ones(2),
            },
        },
    ).adjust_material(dryrock_reference_material, pressure)
    snapshot.assert_match(d)


def test_pressure_dependency_no_pressure_exp_fit(
    snapshot,
    dryrock_reference_material,
    overburden_pressure,
    reference_pressure,
    truncate_pressure,
):
    presrock = 1.0
    pressure = Pressure(
        overburden=overburden_pressure,
        max_effective=truncate_pressure,
        reference=reference_pressure,
        rock=presrock,
        fluid=presrock,
    )
    d = PressureDependency(
        model={
            "type": "expfit",
            "coefficients": {
                "density": np.ones(3),
                "primary_velocity": np.ones(3),
                "secondary_velocity": np.ones(3),
            },
        },
    ).adjust_material(dryrock_reference_material, pressure)
    snapshot.assert_match(d)


def test_pressure_case_max_pressure():
    pressure = Pressure(overburden=30, fluid=10, reference=20, max_effective=11)
    assert_similar_material(
        PressureDependency(
            model={
                "type": "logfit",
                "coefficients": {
                    "density": [0.0, 1.0],
                    "bulk_modulus": [0.0, 1.0],
                    "shear_modulus": [0.0, 1.0],
                },
            },
        ).adjust_material(
            Material(bulk_modulus=1.0, shear_modulus=1.0, density=1.0), pressure
        ),
        Material(
            bulk_modulus=np.log10(11),
            shear_modulus=np.log10(11),
            density=np.log10(11),
        ),
    )


def test_pressure_invalid_reference():
    pressure = Pressure(reference=19, overburden=20, fluid=9)
    with pytest.raises(ValueError, match="zero for reference"):
        PressureDependency(
            model={
                "type": "logfit",
                "coefficients": {
                    "density": [0.0, 1.0],
                    "bulk_modulus": [0.0, 1.0],
                    "shear_modulus": [0.0, 1.0],
                },
            },
        ).adjust_material(
            Material(bulk_modulus=1.0, shear_modulus=1.0, density=1.0), pressure
        ),


def test_pressure_invalid_powerfit_material():
    pressure = Pressure(reference=0.0, overburden=0.0, fluid=np.sqrt(3.0 / 4.0) - 10.0)
    with pytest.raises(ValueError, match="valid dry rock"):
        PressureDependency(
            model={
                "type": "powerfit",
                "coefficients": {
                    "density": [1.0, 1.0],
                    "bulk_modulus": [1.0, 1.0],
                    "vp_over_vs": [1.0, 1.0],
                },
            },
        ).adjust_material(
            Material(primary_velocity=10.0, secondary_velocity=1.0, density=1.0),
            pressure,
        ),


def test_pressure_raises_negative_reference_effective_pressure():
    with pytest.raises(ValueError, match="reference pressure"):
        Pressure(
            overburden=20.0e6,
            reference=30.0e6,
            fluid=15.0e6,
        )


def test_pressure_raises_negative_fluid_effective_pressure():
    with pytest.raises(ValueError, match="fluid pressure"):
        Pressure(
            overburden=50.0e6,
            reference=30.0e6,
            fluid=70.0e6,
        )


def test_pressure_raises_negative_rock_effective_pressure():
    with pytest.raises(ValueError, match="rock pressure"):
        Pressure(
            overburden=50.0e6,
            reference=30.0e6,
            fluid=30.0e6,
            rock=65.0e6,
        )
