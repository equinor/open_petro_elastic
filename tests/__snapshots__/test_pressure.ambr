# serializer version: 1
# name: test_pressure_dependency_no_pressure_exp_fit
  Material(shear_modulus=1.0, bulk_modulus=0.9999999999999998, primary_velocity=1.5275252316519465, secondary_velocity=1.0, density=1.0)
# ---
# name: test_pressure_dependency_no_pressure_poly_fit
  Material(shear_modulus=1.0, bulk_modulus=0.9999999999999998, primary_velocity=1.5275252316519465, secondary_velocity=1.0, density=1.0)
# ---
# name: test_pressure_dependency_no_pressure_power_fit
  Material(shear_modulus=1.0, bulk_modulus=0.9999999999999998, primary_velocity=1.5275252316519465, secondary_velocity=1.0, density=1.0)
# ---
